control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go

```diff
@@ -262,34 +274,54 @@ func getOwnerPrefixFromPod(pod *corev1.Pod) string {
 	return ""
 }
 
-// registerService creates a Consul service registration from the provided Kuberetes service and endpoint information.
-func (r *Controller) registerService(ctx context.Context, resourceClient pbresource.ResourceServiceClient, k8sService corev1.Service, consulSvc *pbcatalog.Service) error {
-	consulSvcResource := r.getServiceResource(
-		consulSvc,
+// ensureService upserts a Consul service registration from the provided service information if a matching service
+// resource does not already exist in Consul.
+func (r *Controller) ensureService(ctx context.Context, resourceClient pbresource.ResourceServiceClient, k8sService corev1.Service, consulSvc *pbcatalog.Service) error {
+	id := getServiceID(
 		k8sService.Name, // Consul and Kubernetes service name will always match
 		r.getConsulNamespace(k8sService.Namespace),
-		r.getConsulPartition(),
-		getServiceMeta(k8sService),
-	)
+		r.getConsulPartition())
+
+	// We use the locally-created Resource and ID (without Uid and Version)
+	// when writing so that it behaves as an upsert rather than CAS.
+	consulSvcResource := &pbresource.Resource{
+		Id:       id,
+		Data:     inject.ToProtoAny(consulSvc),
```

Comment:
I think you need to change the underlying `inject.ToProtoAny` code to not use `any.New` in favor of `any.MarshalFrom` so you can set `Deterministic=true` which should up the chances of actually debouncing.

Reply:
Ah, TIL `Deterministic=true`, thanks for catching this! I checked some other docs and it _looked_ like marshalling was default deterministic within a given version of the language proto impl - but this does looks necessary here.

Reply:
Given `inject.ToProtoAny` is general-purpose and forces sorting, I opted for a direct marshal just for this case, but I think this should work correctly now.